### PR TITLE
added gonda, a gRPC inrospection demo

### PR DESCRIPTION
**Pull request by matholum**
This is a demo showing just a simple password flow but with introspection on a gRPC service/api that's being consumed by a console application. The server is still utilizing rest for the auth calls but everything else is gRPC.

I was wanting to play around with this and saw that there have been several questions about this in the past. I figured it wouldn't hurt to submit a demo to help others curious about this scenario.

### CODE REVIEW

**Comment by kevinchalet** [comment]
Since we already have a sample demonstrating how to use introspection, let's reduce the size of this sample - to make it easier to maintain - by removing the API project and moving it to a single API+server project.

**Comment by matholum** [comment]
@kevinchalet - I thought it was gong to be easy to remove the introspection and extra API project but I ran into issues on mac due to it not supporting TLS on HTTP2 and then trying to mix it w/ HTTPS for the auth endpoints. I may have gotten passed that part but I'm now getting a 401 on the client and the following error on the server:

```
The response was successfully returned as a challenge response: {
        "error": "invalid_token",
        "error_description": "The issuer associated to the specified token is not valid.",
        "error_uri": "https://documentation.openiddict.com/errors/ID2088"
      }
  ```
  
  Any ideas what I'm doing wrong?

**Comment by kevinchalet** [comment]
@matholum the problem is that your API can be reached using two different hosts. Since the local server integration package uses the "Host" header of API requests to determine what the local issuer is, you end up with a token issued by host A and an API request sent to host B, which cannot work. If you can't eliminate one of the two hosts, you'll need to override the issuers in the validation options using `options.Configure(o => o.TokenValidationParameters.ValidIssuers = ...)`.

**Comment by matholum** [comment]
I didn't see anything for `TokenValidationParameters.ValidIssuers` but there was a `SetIssuer(...)` option and that did the trick. Thank you ^_^

File: samples/Gonda/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+obj
```
Comment (kevinchalet) [inline samples/Gonda/.gitignore:1]: Is this file needed?

File: .gitignore [hunk 2]
```diff
@@ -1,6 +1,8 @@
 ## Ignore Visual Studio temporary files, build results, and 
 ## files generated by popular Visual Studio add-ons.
 
+.idea
```
Comment (kevinchalet) [inline .gitignore:4]: Care to explain why it's needed?

File: samples/Gonda/Gonda.Client/Gonda.Client.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Google.Protobuf" />
+    <PackageReference Include="Grpc.Net.Client" />
+    <PackageReference Include="Grpc.Tools">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Newtonsoft.Json" />
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Client/Gonda.Client.csproj:19]: Instead of using JSON.NET, you may want to use the IdentityModel client, which should simply things a bit (it's already used in another client sample so feel free to take a look)

File: samples/Gonda/Gonda.Client/Models/TokensResponse.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+using Newtonsoft.Json;
+
+namespace Gonda.Client.Models
+{
+    public class TokensResponse
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Client/Models/TokensResponse.cs:5]: Once you move to the IM client, this class shouldn't be necessary.

File: samples/Gonda/Gonda.Client/Program.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            var request = new
+            {
+                Email = "test@email.com",
+                Password = "Gonda~1",
+                ConfirmPassword = "Gonda~1"
+            };
+            
+            var content = JsonConvert.SerializeObject(request);
+
+            var response = await authClient.PostAsync("Account/register",
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Client/Program.cs:119]: You may want to use the new HTTP JSON APIs (w/ `JsonContent`) instead of using JSON.NET + `StringContent`.

File: samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                    });
+
+                    return Forbid(properties, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
+                }
+
+                // Create a new ClaimsPrincipal containing the claims that
+                // will be used to create an id_token, a token or a code.
+                var principal = await _signInManager.CreateUserPrincipalAsync(user);
+
+                var scopes = request.GetScopes();
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs:75]: Looks like an unused variable?

File: samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                
+                principal.SetScopes(request.GetScopes());
+                // principal.SetResources(await _scopeManager.ListResourcesAsync(scopes).ToListAsync());
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs:78]: Any particular reason for commenting this call?

File: samples/Gonda/Gonda.Server/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                // Register the OpenIddict validation components.
+                .AddValidation(options =>
+                {
+                    // Import the configuration from the local OpenIddict server instance.
+                    options.UseLocalServer();
+
+                    // Register the ASP.NET Core host.
+                    options.UseAspNetCore();
+
+                    options.SetIssuer("https://localhost:5001");
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Server/Startup.cs:99]: May be worth adding a comment to indicate why it's needed in this sample.

File: samples/Zirku/Zirku.Server/Startup.cs [hunk 9]
```diff
@@ -24,7 +24,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddDbContext<ApplicationDbContext>(options =>
             {
                 // Configure the context to use Microsoft SQL Server.
-                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection"));
+                options.UseInMemoryDatabase(nameof(ApplicationDbContext));
```
Comment (kevinchalet) [inline samples/Zirku/Zirku.Server/Startup.cs:27]: Please remove any unrelated change made to the Zirku sample.

File: samples/Gonda/Gonda.Server/Services/GreeterService.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.Extensions.Logging;
+using OpenIddict.Validation.AspNetCore;
+
+namespace Gonda.Api1
```
Comment (kevinchalet) [inline samples/Gonda/Gonda.Server/Services/GreeterService.cs:11]: Now that the gRPC service and the authorization server are part of the same app, you may want to change all references to the `Gonda.Api1` namespace to something like `Gondo.Server.Services`.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your PR!

File: .gitignore (hunk 2 above)
Comment (matholum) [reply .gitignore:4]: Not required but I'm using Rider as my IDE and wasn't wanting to commit the files it added. I thought adding it here would help any others using Rider too

File: samples/Gonda/.gitignore (hunk 1 above)
Comment (matholum) [reply samples/Gonda/.gitignore:1]: since there is the root level gitignore this can be removed...

File: samples/Gonda/Gonda.Server/Services/GreeterService.cs (hunk 10 above)
Comment (matholum) [reply samples/Gonda/Gonda.Server/Services/GreeterService.cs:11]: Good call

File: samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs (hunk 6 above)
Comment (matholum) [reply samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs:75]: This is due to the commented line below...

File: samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs (hunk 7 above)
Comment (matholum) [reply samples/Gonda/Gonda.Server/Controllers/AuthorizationController.cs:78]: I think I'd removed this while testing something. I just uncommented it and the demo is still working

File: samples/Gonda/Gonda.Server/Gonda.Server.csproj [hunk 11]
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0</TargetFramework>
+    <RootNamespace>Gonda.Server</RootNamespace>
```
[the remark this answers is not in the document]
Comment (matholum) [reply samples/Gonda/Gonda.Server/Gonda.Server.csproj:5]: I think this is just added by default... I removed it and everything seems fine

File: samples/Gonda/Gonda.Client/Gonda.Client.csproj (hunk 3 above)
Comment (matholum) [reply samples/Gonda/Gonda.Client/Gonda.Client.csproj:19]: It doesn't seem like `IdentityModel.OidcClient` supports the Password flow... it's wanting me to add an AuthorizeUrl 🙃

**Comment by matholum** [comment]
@kevinchalet - Sorry I went MIA again for a bit... with the exception of the IM Client I think I've addressed all your comments

File: samples/Gonda/Gonda.Client/Gonda.Client.csproj (hunk 3 above)
Comment (kevinchalet) [reply samples/Gonda/Gonda.Client/Gonda.Client.csproj]: Woops, sorry, I should have been clearer: the method you're looking for is in the `IdentityModel` package, under the `TokenClient` class: https://github.com/IdentityModel/IdentityModel/blob/main/src/Client/TokenClient.cs#L110 😃

File: samples/Gonda/Gonda.Client/Models/TokensResponse.cs (hunk 4 above)
Comment (matholum) [reply samples/Gonda/Gonda.Client/Models/TokensResponse.cs]: removed ^_^

File: samples/Gonda/Gonda.Client/Gonda.Client.csproj (hunk 3 above)
Comment (matholum) [reply samples/Gonda/Gonda.Client/Gonda.Client.csproj]: aha! That one makes more sense... lol. ^_^

**Comment by kevinchalet** [comment]
I tried locally but sadly, I'm getting a weird error from the gRPC tooling that prevents the server project from building:

```
C:\Users\Kévin Chalet\.nuget\packages\grpc.tools\2.39.1\build\_protobuf\Google.Protobuf.Tools.targets(280,5): error MSB6006: "C:\Users\Kévin Chalet\.nuget\packages\grpc.tools\2.39.1\tools\windows_x64\protoc.exe" exited with code -1073741819.
```

I suspect it may be caused by the diacritic or space in my user folder (and it's not something that seems to have been fixed in the most recent gRPC tools package 😢)

**Comment by matholum** [comment]
@kevinchalet - I think it's the accented letter but I think this work around may help you

https://github.com/grpc/grpc/issues/17995#issuecomment-607159949
